README.md
# dbunit

DbUnit，一款关系型数据库单元测试工具。

仿造Java版本的DbUnit，其在单元测试setup中备份需要的数据库表，导入准备好的测试数据，在完成单元测试后在TearDown阶段使用之前备份的文件恢复数据库表原始内容。

目前的版本是 v 0.2

v 0.2版本的特点为：

- 包名改为 dbunit。

## 安装方式

使用pip工具进行安装，如下：

```command
pip install dbunit
```

由于项目处于早期阶段，更新较多，请经常使用如下命令进行版本更新：

```command
pip install dbunit --upgrade
```

## 使用示例

```python
import unittest
from dbunit import DbUnit
import dbunit.utils as utils


class TestDemo(unittest.TestCase):

    def setUp(self):
        self.db_cfg = {
            'server_url': 'rm-bp13oao7f763scs44yo.mysql.rds.aliyuncs.com',
            'server_port': 3306,
            'db_name': 'fdata_dev',
            'user_id': 'dev',
            'user_password': 'dev_61875707',
        }
        self.dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=cfg))
        self.dbu.add_table('table_a', 'test_file_a.csv')
        self.dbu.add_table('table_b')
        self.dbu.backup()
        self.dbu.load_data()

    def tearDown(self):
        self.dbu.reload()

    def test_foo(self):
        # 业务测试
        pass

```


doc/changelog.txt
Changlog

# v 0.1 

- 首次发布
- 支持mysql数据库

# v 0.2

- chg: 包名改为 dbunit


test/test_hi.py
# import unittest
from wdbd.dbunit import DbUnit
# import wdbd.dbunit.utils as utils


if __name__ == "__main__":
    # # dbu = DbUnit()
    # cfg = {
    #     'server_url': 'rm-bp13oao7f763scs44yo.mysql.rds.aliyuncs.com',
    #     'server_port': 3306,
    #     'db_name': 'fdata_dev',
    #     'user_id': 'dev',
    #     'user_password': 'dev_61875707',
    # }
    # dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=cfg), bak_dir=r'c:/temp/dbunit_bak/')
    # print(dbu)
    # print(dbu._db_engine)
    # print(dbu._backup_dir)

    # func:
    # dbu.export_data_file(db_engine=utils.get_conn_engine(cfg), table_name='dw_worker_status', file_path=r'c:/temp/dw_worker_status.csv')

    # # dbu.add_table('dw_worker_status', testdata_file=r'c:/temp/dw_worker_status.csv')
    # dbu.add_table('dw_worker_status')
    # dbu.backup()
    # dbu.load_data()

    # dbu.reload()

    try:
        dbu = DbUnit()
    except Exception:
        print('io err')

    pass


test/test_dbunit.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
'''
@File    :   test_dbunit.py
@Time    :   2020/02/28 18:38:39
@Author  :   Jeffrey Wang
@Version :   1.0
@Contact :   shwangjj@163.com
@Desc    :   Dbunit 单元测试

针对Dbunit模块的功能测试

本测试无需已有数据库表支持，会自动create/drop需要的表

测试内容：
1. 测试参数初始化，如临时文件夹路径是否生成，logger是否生成等
2. backup功能测试
3. load_data功能测试
4. reload功能测试

'''
import unittest
from wdbd.dbunit import DbUnit
import wdbd.dbunit.utils as utils
import os
import tempfile


class Test_DbUnit(unittest.TestCase):
    """测试DbUnit组件
    """

    def setUp(self):
        self.db_cfg = {
            'server_url': 'rm-bp13oao7f763scs44yo.mysql.rds.aliyuncs.com',
            'server_port': 3306,
            'db_name': 'fdata_dev',
            'user_id': 'dev',
            'user_password': 'dev_61875707',
        }
        try:
            # 建立测试用临时表
            session = utils.get_session(utils.get_conn_engine(self.db_cfg))
            sql = "create table dbu_test_a (`name` VARCHAR(20) NOT NULL, PRIMARY KEY (`name`))"
            session.execute(sql)
            sql = "create table dbu_test_b (`id` VARCHAR(20) NOT NULL, PRIMARY KEY (`id`))"
            session.execute(sql)
        finally:
            session.close()
        return super().setUp()

    def tearDown(self):
        try:
            # Drop测试用临时表
            session = utils.get_session(utils.get_conn_engine(self.db_cfg))
            sql = "drop table dbu_test_a"
            session.execute(sql)
            sql = "drop table dbu_test_b"
            session.execute(sql)
        finally:
            session.close()
        return super().tearDown()

    def test_dbunit_init(self):
        """
        测试Dbunit对象的初始化
        """
        dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=self.db_cfg))
        self.assertDictEqual({}, dbu.back_files)
        self.assertDictEqual({}, dbu.test_data)
        # 默认的临时文件夹，应该是操作系统临时目录下的\dbtest_temp\
        bak_dir = tempfile.gettempdir() + r'\dbtest_temp'
        self.assertEqual(bak_dir, dbu._backup_dir)

        # 测试初始化参数为空，无法连接数据库的情况
        try:
            dbu = DbUnit()
            self.fail()
        except IOError:
            pass
        else:
            self.fail()

    def test_backup(self):
        """测试 backup 函数功能
        """
        dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=self.db_cfg))
        dbu.add_table('dbu_test_a', r'test\dbu_test_a.csv')
        dbu.add_table('dbu_test_b')     # 无测试数据情况
        dbu.backup()

        # 检查登记表情况
        bak_files = {
            'dbu_test_a': dbu._backup_dir + r'\dbu_test_a.csv',
            'dbu_test_b': dbu._backup_dir + r'\dbu_test_b.csv',
        }
        self.assertDictEqual(bak_files, dbu.back_files)

        # 检查备份文件是否生成
        self.assertTrue(os.path.exists(dbu._backup_dir + r'\dbu_test_a.csv'))
        self.assertTrue(os.path.exists(dbu._backup_dir + r'\dbu_test_b.csv'))

        # 测试表不存在的情况
        try:
            dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=self.db_cfg))
            dbu.add_table('xxxx')
            dbu.backup()
            self.fail()
        except Exception as err:
            self.assertIsNotNone(err)

    def test_load_data(self):
        """测试 load_data 函数功能
        """
        dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=self.db_cfg))
        dbu.add_table('dbu_test_a', r'test\dbu_test_a.csv')
        dbu.add_table('dbu_test_b')     # 无测试数据情况
        dbu.backup()
        dbu.load_data()

        # 检查 测试数据表情况
        test_data = {
            'dbu_test_a': r'test\dbu_test_a.csv',
            'dbu_test_b': None,
        }
        self.assertDictEqual(test_data, dbu.test_data)

        # 检查数据库表内容是否同测试数据一致
        try:
            session = utils.get_session(utils.get_conn_engine(self.db_cfg))
            # 检查表A：
            sql = "select * from dbu_test_a"
            cursor = session.execute(sql)
            records = cursor.fetchall()
            data = [('张三',), ('李四',)]
            self.assertListEqual(data, records)
            cursor.close()
            # 检查表B：
            sql = "select * from dbu_test_b"
            cursor = session.execute(sql)
            records = cursor.fetchall()
            self.assertListEqual([], records)
            cursor.close()
        finally:
            session.close()

    def test_reload(self):
        """测试 reload 函数功能
        """
        dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=self.db_cfg))
        dbu.add_table('dbu_test_a', r'test\dbu_test_a.csv')
        dbu.add_table('dbu_test_b')     # 无测试数据情况
        dbu.backup()
        dbu.load_data()
        dbu.reload()

        # 检查数据库表内容是否恢复最初情况？
        try:
            session = utils.get_session(utils.get_conn_engine(self.db_cfg))
            # 检查表A：
            sql = "select * from dbu_test_a"
            cursor = session.execute(sql)
            records = cursor.fetchall()
            self.assertListEqual([], records)
            cursor.close()
            # 检查表B：
            sql = "select * from dbu_test_b"
            cursor = session.execute(sql)
            records = cursor.fetchall()
            self.assertListEqual([], records)
            cursor.close()
        finally:
            session.close()


test_dbunit/test_hi.py
# import unittest
# from dbunit import DbUnit
# import wdbd.dbunit.utils as utils


if __name__ == "__main__":
    # # dbu = DbUnit()
    # cfg = {
    #     'server_url': 'rm-bp13oao7f763scs44yo.mysql.rds.aliyuncs.com',
    #     'server_port': 3306,
    #     'db_name': 'fdata_dev',
    #     'user_id': 'dev',
    #     'user_password': 'dev_61875707',
    # }
    # dbu = DbUnit(db_engine=utils.get_conn_engine(db_config=cfg), bak_dir=r'c:/temp/dbunit_bak/')
    # print(dbu)
    # print(dbu._db_engine)
    # print(dbu._backup_dir)

    # func:
    # dbu.export_data_file(db_engine=utils.get_conn_engine(cfg), table_name='dw_worker_status', file_path=r'c:/temp/dw_worker_status.csv')

    # # dbu.add_table('dw_worker_status', testdata_file=r'c:/temp/dw_worker_status.csv')
    # dbu.add_table('dw_worker_status')
    # dbu.backup()
    # dbu.load_data()

    # dbu.reload()

    # try:
    #     dbu = DbUnit()
    # except Exception:
    #     print('io err')

    print('Hi')

    pass


dbunit/dbunit.py


import os
import pandas as pd
import tempfile
import dbunit.utils as utils


class DbUnitImpl:
    """
    关系型数据库 单元测试工具
    """

    def __init__(self, db_engine=None, db_config=None, logger=None, bak_dir=None):
        self.back_files = {}    # 备份文件注册表，表名:备份文件
        self.test_data = {}     # 测试文件注册表，表名:测试文件
        if not logger:
            self._logger = utils.get_logger()
        else:
            self._logger = logger       # 日志

        # 数据库连接
        if not db_engine:
            self._db_engine = utils.get_conn_engine(db_config)
        else:
            self._db_engine = db_engine
        if self._db_engine is None:
            self._logger.error('无数据库连接，无法进行单元测试！')
            raise IOError('无数据库连接，无法进行单元测试！')

        if bak_dir is None:
            # 默认备份文件存放在系统临时文件夹下
            self._backup_dir = tempfile.gettempdir() + r'\dbtest_temp'
        else:
            self._backup_dir = bak_dir
        if not os.path.exists(self._backup_dir):
            os.mkdir(self._backup_dir)

    @staticmethod
    def export_data_file(db_engine, table_name, file_path):
        # 工具函数，数据库表导出备份文件
        df = pd.read_sql_table(table_name=table_name.lower(), con=db_engine)
        if df is not None:
            df.to_csv(file_path, index=False, encoding='utf-8')
            print('export {0} ==> {1} ok'.format(table_name, file_path))

    def _log(self, msg):
        # 记录日志
        if self._logger:
            self._logger.debug("【dbunit】" + msg)
        # else:
        #     print("【dbunit】" + msg)

    def add_table(self, tablename, testdata_file=None):
        # 添加 需要测试的表名测试用文件
        # testdata_file = None，说明改表仅需要清空，不要导入测试文件
        self.back_files[tablename] = None
        self.test_data[tablename] = testdata_file

    def backup(self):
        """备份数据表内容到临时文件
        """
        self._log('开始备份 ... ')
        # 建立临时文件夹
        if not os.path.exists(self._backup_dir):
            os.mkdir(self._backup_dir)

        try:
            # engine = utils.get_conn_engine()
            engine = self._db_engine
            session = utils.get_session(engine)

            # 所有表导出临时备份文件
            count_all = len(self.back_files)
            idx = 1
            for table_name in self.back_files:
                df = pd.read_sql_table(table_name=table_name.lower(), con=engine)
                # print(df)

                # if df is not None and not df.empty:
                if df is not None:
                    # 生成备份文件
                    bak_file = self._backup_dir + "\\" + table_name + '.csv'
                    df.to_csv(bak_file, index=False, encoding='utf-8', chunksize=10000)
                    self._log('[{0}/{1}] 表 {2} 备份到 {3} '.format(idx, count_all, table_name, bak_file))
                    self.back_files[table_name] = bak_file    # 登记
                else:
                    # 原始表无数据
                    self._log('[{0}/{1}] 表 {2} 中无数据 ... '.format(idx, count_all, table_name))
                idx += 1
            # self._log('back_files {0}'.format(self.back_files))
            # self._log('test_data {0}'.format(self.test_data))
            self._log('全部{0}个表备份完成！'.format(count_all))
        except Exception as err:
            self._log('备份原始数据时出现问题，' + str(err))
            raise Exception('备份原始数据时出现问题，' + str(err))
        finally:
            session.close()

    def load_data(self):
        """导入测试数据
        """
        self._log('导入测试数据 ... ')

        try:
            # engine = utils.get_conn_engine()
            engine = self._db_engine
            session = utils.get_session(engine)

            count_all = len(self.test_data)
            idx = 1
            for table_name in self.test_data:
                data_file = self.test_data[table_name]

                # 删除原始数据
                sql = "truncate table " + table_name
                session.execute(sql)

                if data_file is None:
                    # 无测试文件
                    self._log('[{0}/{1}] 表 {2} 无 测试文件 '.format(idx, count_all, table_name))
                elif os.path.exists(data_file):
                    # 导入测试数据
                    df = pd.read_csv(data_file)
                    # 导入数据库表
                    df.to_sql(name=table_name, con=engine, if_exists='append', index=False)
                    self._log('[{0}/{1}] 表 {2} 导入测试数据  [{3}] '.format(idx, count_all, table_name, df.shape[0]))
                else:
                    # 测试文件不存在
                    self._log('[{0}/{1}] 表 {2} 测试文件不存在 [{3}] '.format(idx, count_all, table_name, data_file))
                idx += 1

            self._log('导入测试数据完成！')
        except Exception as err:
            self._log('导入测试数据时出现问题，' + str(err))
        finally:
            session.close()

    def reload(self):
        """恢复原始数据
        """
        self._log('恢复数据 ... ')

        try:
            # engine = utils.get_conn_engine()
            engine = self._db_engine
            session = utils.get_session(engine)

            count_all = len(self.back_files)
            idx = 1
            for table_name in self.back_files:
                data_file = self.back_files[table_name]

                # 删除原始数据
                sql = "truncate table " + table_name
                session.execute(sql)

                if not data_file:
                    self._log('[{0}/{1}] 表 {2} 没有备份，无法恢复'.format(idx, count_all, table_name))
                elif os.path.exists(data_file):
                    # 导入测试数据
                    df = pd.read_csv(data_file)
                    # 导入数据库表
                    df.to_sql(name=table_name, con=engine, if_exists='append', index=False)
                    self._log('[{0}/{1}] 表 {2} 恢复 [{3}] '.format(idx, count_all, table_name, df.shape[0]))
                else:
                    # 测试文件不存在
                    self._log('[{0}/{1}] 表 {2} 备份文件不存在，无法恢复 [{3}] '.format(idx, count_all, table_name, data_file))
                idx += 1

            # TODO 删除建立的临时文件夹
            self._log('恢复数据完成！')
        except Exception as err:
            self._log('恢复数据时出现问题，' + str(err))
        finally:
            session.close()


dbunit/utils.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-
'''
@File    :   utils.py
@Time    :   2020/04/13 20:53:20
@Author  :   Jeffrey Wang
@Version :   1.0
@Contact :   shwangjj@163.com
@Desc    :   工具模块
'''
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
import logging.config


def get_logger():
    """
    返回一个日志器
    :rtype:
    :param config_file_path: str 配置文件路径
    :return: logger日志器
    """
    config_file_path = r'dbunit/log.cfg'

    logging.config.fileConfig(config_file_path)
    return logging.getLogger('dbunit')


def get_conn_engine(db_config):
    """返回数据库连接引擎

    demo for db_config:
    {
        'server_url': '127.0.0.1',
        'server_port': 3306,
        'db_name': 'fdata_dev',
        'user_id': 'root',
        'user_password': '123456',
        'echo': false,
        'max_overflow': 0,
        'pool_size': 5,
        'pool_timeout': 10,
        'pool_recycle': -1,
    }

    Arguments:
        db_config {dict} -- mysql数据库连接配置

    Returns:
        [engine] -- 数据库连接引擎
    """
    log = get_logger()
    if not db_config:
        log.error('数据库配置为空，无法建立数据库连接')
        return None

    # mysql+mysqlconnector://dev:dev_61875707@rm-bp13oao7f763scs44yo.mysql.rds.aliyuncs.com:3306/fdata_dev
    try:
        conn_str = 'mysql+mysqlconnector://{user_id}:{user_password}@{server_url}:{server_port}/{db_name}'.format(
            server_url=db_config.get('server_url'),
            user_id=db_config.get('user_id'),
            user_password=db_config.get('user_password'),
            server_port=db_config.get('server_port'),
            db_name=db_config.get('db_name'))
        # engine = create_engine(conn_str, echo=True)
        engine = create_engine(conn_str, echo=db_config.get('echo', False),
                               max_overflow=db_config.get(
                                   'max_overflow', 0),  # 超过连接池大小外最多创建的连接
                               pool_size=db_config.get(
                                   'pool_size', 5),  # 连接池大小
                               pool_timeout=db_config.get(
                                   'pool_timeout', 10),  # 池中没有线程最多等待的时间，否则报错
                               # 多久之后对线程池中的线程进行一次连接的回收（重置）
                               pool_recycle=db_config.get('pool_recycle', -1)
                               )
        return engine
    except Exception as err:
        log.error("数据库连接失败，" + str(err))
        return None


def get_session(engine):
    """新建并返回数据库会话（session）

    如果参数提供engine则使用之，否则新建

    Keyword Arguments:
        engine {sqlalchemy.engine} -- 如果参数提供engine则使用之，否则新建 (default: {None})

    Returns:
        [sqlalchemy.session] -- 数据库会话
    """
    log = get_logger()
    try:
        SessionFactory = sessionmaker(engine)
        db_session = SessionFactory()
        return db_session
    except Exception as err:
        log.error(str(err))
        return None


if __name__ == "__main__":
    cfg = {
        'server_url': 'rm-bp13oao7f763scs44yo.mysql.rds.aliyuncs.com',
        'server_port': 3306,
        'db_name': 'fdata_dev',
        'user_id': 'dev',
        'user_password': 'dev_61875707',
    }
    e = get_conn_engine(db_config=cfg)
    session = get_session(engine=e)
    print(e)
    print(session)
